src/centrality/truss.cpp

```diff
@@ -0,0 +1,254 @@
+
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc.,  51 Franklin Street, Fifth Floor, Boston, MA
+  02110-1301 USA
+
+*/
+
+#include "igraph_adjlist.h"
+#include "igraph_community.h"
+#include "igraph_error.h"
+#include "igraph_interface.h"
+#include "igraph_motifs.h"
+
+#include <iostream>
+#include <vector>
+#include <unordered_set>
+
+using namespace std;
+
+/* helper functions */
+static void igraph_truss_i_unpack(const igraph_vector_int_t *tri, igraph_vector_t *unpacked_tri);
+static void igraph_truss_i_compute_support(const igraph_vector_t *eid, igraph_vector_int_t *support);
+static int igraph_i_trussness(const igraph_t *graph, igraph_vector_int_t *support,
+  igraph_vector_int_t *truss);
+
+/**
+ * \function trussness
+ * \brief Find the "trussness" for every edge in the graph. This
+ * indicates the highest k-truss that the edge occurs in.
+ *
+  * </para><para>
+ * A k-truss is a subgraph in which every edge occurs in at least k-2 triangles in the subgraph.
+ *
+ * </para><para>
+ * This function returns the highest k for each edge. If you are interested in
+ * a particular k-truss subgraph, you can subset the graph using to those eids
+ * which are >= k because each k-truss is a subgraph of a (k–1)-truss
+ * (thus to get * all 4-trusses, take k >= 4 because the 5-trusses, 6-trusses, etc need to be included).
+ *
+ * </para><para>The current implementation of this function iteratively
+ * decrements support of each edge using O(|E|) space and O(|E|^1.5) time.
+ *
+ * \param graph The input graph.
+ * \param truss Pointer to initialized vector of truss values that will
+ * indicate the highest k-truss each edge occurs in. It will be resized as
+ * needed.
+ * \return Error code.
+ *
+ * Time complexity: It should be O(|E|^1.5). See Algorithm 2 in:
+ * Wang, Jia, and James Cheng. "Truss decomposition in massive networks."
+ * Proceedings of the VLDB Endowment 5.9 (2012): 812-823.
+*
+ */
+int igraph_trussness(const igraph_t* graph, igraph_vector_int_t* truss){
+  igraph_vector_int_t triangles, support;
+  igraph_vector_t eid, unpacked_triangles;
+
+  /* Manage the stack to make it memory safe */
+  IGRAPH_VECTOR_INT_INIT_FINALLY(&support, igraph_ecount(graph));
+  IGRAPH_VECTOR_INIT_FINALLY(&eid, 0);
+  IGRAPH_VECTOR_INIT_FINALLY(&unpacked_triangles, 0);
+  IGRAPH_VECTOR_INT_INIT_FINALLY(&triangles, 0);  // will be resized
```

Comment:
Add a comment here that mentions that `triangles` must be the last one as it will be the first vector to be destroyed and popped off the `FINALLY` stack later. In case someone feels the urge to reorder the variable initializations in alphabetic order ;)